### ISSUE TITLE
Delete operation - Different behavior in CypherQuery execution vs API

### ISSUE THREAD

**Issue by bHacklv** [neo4j/neo4j#6323]
## Description of the bug:

Within 1st transaction we create (node1) node, (node2) node, (node1)-[rel]->(node2) relationship.

Within 2nd transaction we want to Delete parent node, child node and then relationship (or do that in any order). 

This is method call order:
1. Node#delete() **(on node1)**
2. Node#delete() **(on node2)**
3. Relationship#delete() **(on (node1)-[rel]->(node2))**

However, if we execute these cyphers:
1. GraphDatabaseService#execute("match n where n.uuid = '" + parentNodeId + "' delete n") **(deletes node1)**
2. GraphDatabaseService#execute("match n where n.uuid = '" + childNodeId+ "' delete n") **(deletes node2)**
3. GraphDatabaseService#execute("match ()-[r]-() where id(r) = " + relId + " return r") **(deletes (node1)-[rel]->(node2) relationship)**

**It throws this exception:**

```
org.neo4j.graphdb.TransactionFailureException: Transaction was marked as successful, but unable to commit transaction so rolled back.
    at org.neo4j.kernel.TopLevelTransaction.close(TopLevelTransaction.java:121)
    at tests.integration.primitives.node.DeleteTest.asd(DeleteTest.java:153)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at tests.integration.util.runner.IntegrationTestRunner.gentlyRunChild(IntegrationTestRunner.java:163)
    at tests.integration.util.runner.IntegrationTestRunner.runChild(IntegrationTestRunner.java:75)
    at tests.integration.util.runner.IntegrationTestRunner.runChild(IntegrationTestRunner.java:42)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
    at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
    at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
    at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: org.neo4j.kernel.api.exceptions.TransactionFailureException: Node record Node[0,used=false,rel=0,prop=-1,labels=Inline(0x0:[]),light] still has relationships
    at org.neo4j.kernel.impl.transaction.state.IntegrityValidator.validateNodeRecord(IntegrityValidator.java:52)
    at org.neo4j.kernel.impl.transaction.state.TransactionRecordState.extractCommands(TransactionRecordState.java:170)
    at org.neo4j.kernel.impl.api.KernelTransactionImplementation.commit(KernelTransactionImplementation.java:539)
    at org.neo4j.kernel.impl.api.KernelTransactionImplementation.close(KernelTransactionImplementation.java:458)
    at org.neo4j.kernel.TopLevelTransaction.close(TopLevelTransaction.java:97)
    ... 30 more
```
## Additional observations:
- Changing cypher query deletion order to: delete node1, rel, node2 has same result (exception is thrown).
- Printing out each execution **getQueryStatistics()** shows that:

```
Nodes deleted: 1
Nodes deleted: 1

```

Last line is empty line for 'delete relationship' query operation (you have a bug in QueryStatistics.toString() method, where **if (result.isEmpty) "Nothing happened" else result** check is always false, since you create **val builder = new StringBuilder("\n")** with 'new-line', which makes it always nonEmpty.
- If we change delete to return for relationship (the rest of the code stay the same), the relationship is returned - **System.out.println(result.next());** will print out **{r=Relationship[0]}**.
- If we have only 2 queries - first deletes parent (or child) node, and second matches relationship and other node and deletes both within one query _"match ()-[r]-(n) where n.uuid = '" + childNodeId + "' delete r, n"_ - will return empty line too, meaning nothing was executed for this operation. Same happens if we swap **r** and **n** after **delete**. And if we return the results for this query, rather than delete - we will see both Node and Relationship objects being returned.
### API

``` java
@Test
    public void deleteNodesAndRelWithApi() throws Exception {
        final DynamicRelationshipType relType = DynamicRelationshipType.withName("CHILD");
        final String parentNodeId;
        final String childNodeId;
        try (final Transaction transaction = database().beginTx()) {
            final Node parentNode = helpers().database().createUniqueNode();
            final Node childNode = helpers().database().createUniqueNode();
            parentNode.createRelationshipTo(childNode, relType);

            parentNodeId = uuid(parentNode);
            childNodeId = uuid(childNode);

            transaction.success();
        }

        try (final Transaction transaction = database().beginTx()) {
            database().findNode(DatabaseHelper.UNIQUE_LABEL, DatabaseHelper.UNIQUE_PROPERTY, parentNodeId).delete();
            database().findNode(DatabaseHelper.UNIQUE_LABEL, DatabaseHelper.UNIQUE_PROPERTY,
                    childNodeId).getRelationships().iterator().next().delete();
            database().findNode(DatabaseHelper.UNIQUE_LABEL, DatabaseHelper.UNIQUE_PROPERTY, childNodeId).delete();
            transaction.success();
        }
    }
```
### Cypher

``` java
@Test
    public void deleteNodesAndRelWithCypherQuery() throws Exception {
        final DynamicRelationshipType relType = DynamicRelationshipType.withName("CHILD");
        final String parentNodeId;
        final String childNodeId;
        final long relId;
        try (final Transaction transaction = database().beginTx()) {
            final Node parentNode = helpers().database().createUniqueNode();
            final Node childNode = helpers().database().createUniqueNode();
            relId = parentNode.createRelationshipTo(childNode, relType).getId();


            parentNodeId = uuid(parentNode);
            childNodeId = uuid(childNode);

            transaction.success();
        }

        try (final Transaction transaction = database().beginTx()) {

            try (Result result = database().execute("match n where n.uuid = '" + parentNodeId + "' delete n")) {
                System.out.println(result.getQueryStatistics());
                while (result.hasNext()) {
                    System.out.println(result.next());
                }
            }
            try (Result result = database().execute("match n where n.uuid = '" + childNodeId + "' delete n")) {
                System.out.println(result.getQueryStatistics());
                while (result.hasNext()) {
                    System.out.println(result.next());
                }
            }

            try (Result result = database().execute("match ()-[r]-() where id(r) = " + relId + " delete r")) {
                System.out.println(result.getQueryStatistics());
                while (result.hasNext()) {
                    System.out.println(result.next());
                }
            }

//            try (Result result = database().execute("match ()-[r]-(n) where n.uuid = '" + childNodeId + "' delete r, n")) {
//                while(result.hasNext()) {
//                    final Map<String, Object> next = result.next();
//                    ((Node) next.get("n")).delete();
//                    ((Relationship) next.get("r")).delete();
//                    System.out.println(next);
//                }
//            }
            transaction.success();
        }
    }
```

(@FylmTM)

**Comment by Mats-SX** [comment]
Hello @bHacklv and thank you for reporting this.

The issue with your fourth (commented-out) query is that you're matching via the child node's uuid to get the relationship. That child node was (recently) deleted, and so the match will not be able to find a node which passes the uuid predicate. Therefore it will also not find a relationship from that node, and subsequently not delete the (dangling) relationship.

If you in your third query replace the `RETURN` for `DELETE`, like so `MATCH ()-[r]-() WHERE id(r) = ... DELETE r`, it should work.

But using a wrapping transaction is not necessary for Cypher, and is slightly convoluted. I would recommend that you merge all three of your queries into one Cypher query accomplishing the same thing in one operation:

```
MATCH (n)
WHERE n.uuid IN [{param1}, {param2}] // or using literals
MATCH ()-[r]->()
WHERE id(r) = {param3}
DELETE n, r // any order here is fine
```

Alternative to using relationship by id-lookup, use the same match from your fourth query to find `r`. In my query above it will work, because no node is deleted until the `DELETE` clause is reached.

The `new StringBuilder("\n")` does look like a bug, though. We'll fix that for 2.3.3.

**Comment by bHacklv** [comment]
Hi @Mats-SX,

Sorry, my 3rd query supposed to have 'delete' instead of 'return' (changed original post) - and no, it doesn't work. I get the same exception stack trace in the end (_...Caused by: org.neo4j.kernel.api.exceptions.TransactionFailureException: Node record Node[0,used=false,rel=0,prop=-1,labels=Inline(0x0:[]),light] still has relationships..._). If I print out the **getQueryStatistics** for **Result result = database().execute("match ()-[r]-() where id(r) = " + relId + " delete r")** it again prints "empty-line", instead of expected **Relationships deleted: 1**.
So there is definitely a bug.

I was trying to demonstrate that 3rd and 4th queries don't work for **delete** but they do work for **return** (even if the node is deleted, i can still match the relationship using that deleted node and return the rel result...).

Again, this is only demonstration of the bug and I have more complex logic behind it, where nodes sometimes have to be deleted before the relationships using cypher (and there I'm matching nodes and relationships by IDs only).

**Comment by FylmTM** [comment]
Comments from me.
If we summarize problem in one simple sentence, then:

> We can't delete relationship in separate Cypher query, in same transaction, if one of the relationship nodes (either start or end node) is deleted before. We can `return` this relationship, but can't `delete`.

**Comment by Mats-SX** [comment]
Hi @bHacklv and @FylmTM 

Yes, you're right, there's an inconsistency here with how Cypher behaves. Semantically, it should be no different using `RETURN` or `DELETE` for deciding if there is a match or not.

What actually is happening here can be shown in the plan descriptions of the queries when substituting `DELETE` and `RETURN`. The former uses the rule planner, which does not have a by-id lookup for relationships, and the latter uses the cost planner, which does not have support for update clauses (in 2.3). What the rule planner tries to do to find the relationship is to find the node and expand, but given that all nodes are deleted, no expansion can be performed. (Try setting `CYPHER planner=rule` for the `RETURN` query, and you'll see it will not be able to find a relationship.)

It seems like you are already able to work around this, but I might as well mention that there's also the `DETACH DELETE` operator (http://neo4j.com/docs/2.3.2/query-delete.html#delete-delete-a-node-with-all-its-relationships) that might be helpful for you. Of course, just deleting the relationship first should also work.

In the upcoming 3.0 release, there will be support for `DELETE` (and other update clauses) in the cost-based planner, and that will make use of the relationship-by-id operator used by your read-version of the third query.

For 2.3, I'll take this to the team for discussion. The use-case is not what is generally recommended for how to use Cypher, which is designed around always leaving the database in a consistent state after each query (which means that your first query would fail if not wrapped in an extra transaction).

**Comment by Mats-SX** [comment]
Hello again @bHacklv and @FylmTM 

So, there's a decent workaround for this specific case that you may use; the `START` clause. If you write your third query like this instead:

```
START r=relationship(relId)
DELETE r
```

Then you should get your expected behaviour. This behaviour is deprecated, however, and could be removed in the future.

We'll not fix the rule planner to be able to use `MATCH` to do lookup-by-id, since we instead have made `DELETE` available to the cost planner in 3.0 (you could check out `3.0.0-M02`, where it should be included, or the upcoming `M03` where full write-support is implemented), and are planning to phase out the rule planner.

The best-practice Cypher recommendation is however to delete the relationship first, then its nodes (or all three in one query).

**Comment by bHacklv** [comment]
@Mats-SX Ok thanks, I've tried using 'start' and it works as expected. Thanks for explanation.
We will use this for now, and wait for 3.0.0 release.
